### PR TITLE
Reduce arguments method buff! at model buffer_update

### PR DESCRIPTION
**Pull request by brunnohenrique**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Refactoring the model `buffer_update.rb` to reduce the arguments in the method `buff!`

## Related Tickets & Documents
https://github.com/forem/forem/issues/3739

## Added tests?

- [ ] yes
- [X] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [X] no documentation needed

### CODE REVIEW

File: app/labor/bufferizer.rb [hunk 1]
```diff
@@ -21,21 +21,22 @@ def satellite_tweet!
       if text.length < 250 && SiteConfig.twitter_hashtag
         text = text.gsub(" #{SiteConfig.twitter_hashtag}", " #{SiteConfig.twitter_hashtag} ##{tag.name}")
       end
-      BufferUpdate.buff!(@article.id, text, tag.buffer_profile_id_code, "twitter", tag.id, @admin_id)
+      BufferUpdate.buff!(@article.id, text, tag.buffer_profile_id_code, { tag_id: tag.id, admin_id: @admin_id })
```
Comment (rhymes) [inline app/labor/bufferizer.rb:24]: It's odd that rubocop didn't flag the parenthesis here as they are redudant 🤔

File: app/models/buffer_update.rb [hunk 2]
```diff
@@ -5,9 +5,11 @@ class BufferUpdate < ApplicationRecord
   validate :validate_body_text_recent_uniqueness, :validate_suggestion_limit
   validates :status, inclusion: { in: %w[pending sent_direct confirmed dismissed] }
 
-  def self.buff!(
-    article_id, text, buffer_profile_id_code, social_service_name = "twitter", tag_id = nil, admin_id = nil
-  )
+  def self.buff!(article_id, text, buffer_profile_id_code, **options)
```
Comment (rhymes) [inline app/models/buffer_update.rb:8]: can we call this `**kwargs` for consistency?

File: app/models/buffer_update.rb [hunk 3]
```diff
@@ -5,9 +5,11 @@ class BufferUpdate < ApplicationRecord
...
+    social_service_name = options.fetch(:social_service_name, "twitter")
+    tag_id = options.fetch(:tag_id, nil)
+    admin_id = options.fetch(:admin_id, nil)
```
Comment (rhymes) [inline app/models/buffer_update.rb:11]: I don't think you need to use fetch here. `kwargs[:tag_id]` is perfectly fine. As it's going to be nil anyway if the value is not given

File: app/labor/bufferizer.rb (hunk 1 above)
Comment (wenderjean) [inline app/labor/bufferizer.rb:24]: The `tag` object itself has been used twice at the same calling, it suggests to me that you could just pass the object itself for the `buff!` method.

File: app/labor/bufferizer.rb [hunk 4]
```diff
@@ -21,21 +21,22 @@ def satellite_tweet!
...
     end
     @article.update(last_buffered: Time.current)
   end
 
   def main_tweet!
-    BufferUpdate.buff!(@article.id, twitter_buffer_text, ApplicationConfig["BUFFER_TWITTER_ID"], "twitter", nil,
-                       @admin_id)
+    BufferUpdate.buff!(@article.id, twitter_buffer_text, ApplicationConfig["BUFFER_TWITTER_ID"],
```
Comment (wenderjean) [inline app/labor/bufferizer.rb:30]: Considering you're using the `buff!` method on different places I believe you could consider something like:

```ruby
class Bufferizer
  ...
  def facebook_post!
    buffer_params = build_buffer_params(text: fb_buffer_text, social_service_name: "facebook")

    BufferUpdate.buff!(buffer_params)
  end

  def build_buffer_params(text:, social_service_name:)
    BufferUpdate::Params.new(
      @article, @tag, @admin, text, social_service_name
    )
  end
  ...
end
```

File: app/labor/bufferizer.rb [hunk 5]
```diff
@@ -21,21 +21,22 @@ def satellite_tweet!
...
+                       { admin_id: @admin_id })
     @article.update(last_buffered: Time.current)
   end
 
   def facebook_post!
-    BufferUpdate.buff!(@article.id, fb_buffer_text, ApplicationConfig["BUFFER_FACEBOOK_ID"], "facebook", nil, @admin_id)
-    BufferUpdate.buff!(@article.id, fb_buffer_text + social_tags, ApplicationConfig["BUFFER_LINKEDIN_ID"], "linkedin",
-                       nil, @admin_id)
+    BufferUpdate.buff!(@article.id, fb_buffer_text, ApplicationConfig["BUFFER_FACEBOOK_ID"],
```
Comment (wenderjean) [inline app/labor/bufferizer.rb:36]: Following my last suggestion, you could create on `ApplicationConfig` an accessor to return that ID using the `social_service_name` and extract it directly inside de `BufferUpdate.buff!`

File: app/models/buffer_update.rb [hunk 6]
```diff
@@ -1,13 +1,24 @@
 class BufferUpdate < ApplicationRecord
+  SOCIAL_SERVICE = {
```
Comment (rhymes) [inline app/models/buffer_update.rb:2]: should this be plural?

File: app/models/buffer_update.rb [hunk 7]
```diff
@@ -54,6 +65,10 @@ def self.twitter_default_text(article)
     ].compact.join.strip
   end
 
+  def self.wich_social_service(service_name)
+    SOCIAL_SERVICE.fetch(service_name)
+  end
```
Comment (rhymes) [inline app/models/buffer_update.rb:70]: side from the typo, I don't think we need the method at all, you can simply call `SOCIAL_SERVICES[social_service_name]` in the `buff!` method

File: app/models/buffer_update.rb (hunk 6 above)
Comment (brunnohenrique) [reply app/models/buffer_update.rb:2]: How about `SOCIAL_MEDIA`? My next commit will change it.

Comment (rhymes) [reply app/models/buffer_update.rb:2]: sure :-)

File: app/labor/bufferizer.rb (hunk 4 above)
Comment (brunnohenrique) [reply app/labor/bufferizer.rb]: I've tried to do a `params_builder` but I think it was too complex. It seems like that:

```rb
  def build_buffer_params(text, social_service_name:, tag:)
    @article.id,
    text,
    Hash.new
      .merge({social_service_name: social_service_name}) if social_service_name
      .merge({buffer_profile_id_code: tag.buffer_profile_id_code}) if tag && tag.buffer_profile_id_code
      .merge({tag_id: tag.id}) if tag
      .merge({admin_id: @admin.id}) if @admin
  end
```

What did you think about it?

**Comment by brunnohenrique** [comment]
@rhymes @wenderjean thanks a lot for the reviews :champagne: 
If there is something else to do, just ping me.

File: app/models/buffer_update.rb [hunk 8]
```diff
@@ -1,11 +1,11 @@
 class BufferUpdate < ApplicationRecord
-  SOCIAL_SERVICE = {
+  SOCIAL_MEDIA = {
     "twitter" => ApplicationConfig["BUFFER_TWITTER_ID"],
     "facebook" => ApplicationConfig["BUFFER_FACEBOOK_ID"],
     "linkedin" => ApplicationConfig["BUFFER_LINKEDIN_ID"]
   }.freeze
 
-  DEFAULT_SOCIAL_SERVICE = "twitter".freeze
+  DEFAULT_SOCIAL_MIDIA = "twitter".freeze
```
Comment (rhymes) [inline app/models/buffer_update.rb:8]: @brunnohenrique there's a small typo here :-)

Comment (brunnohenrique) [reply app/models/buffer_update.rb]: adjusted with `git push -f` to don't to pollute the commits

**Comment by juliannatetreault** [comment]
TIL `**kwargs`! Thanks for taking the time to refactor this - it all looks good to me! 👍 👍

**Pull request merged**